### ISSUE TITLE
TypeScript interface is not working as expected, inside '.vue' files.

### ISSUE THREAD

**Issue by CodeLookBook** [TypeStrong/ts-loader#563]
TypeScript interface doesn't work as expected with '*.vue' files. it doesn't throw any "warning" when provided type is bad. Example form vue.js file:

```
<script lang="ts">

    interface Test {
       w: string;
       h: number;
    }

   // This part of code doesn't work. 
   // It doesn't show any message about 
   // type incompatibility 
   //(propert h must have "number" type).

   let myVar: Test = {
       w: "10",
       h: "26",
   };

   //At the same time this code works correctly in '.ts' files.

   </script>
```

As I understand that's the problem of 'ts-loader'. Is there a way to fix this?

My devDependencies info:

"devDependencies": {
    "babel-plugin-transform-object-rest-spread": "*",
    "cross-env": "^3.2.3",
    "laravel-mix": "0.*",
    "normalize.css": "*",
    "pug": "*",
    "ts-loader": "^2.2.0",
    "typescript": "^2.3.4",
    "vue": "*",
    "vue-resource": "*",
    "vue-router": "*"
  },

**Comment by johnnyreilly** [comment]
I don't think `script lang="ts"` would ever work. How would tsc transpile inline text?

**Comment by johnnyreilly** [comment]
Does your config look like this? 

```
module.exports = {
    entry: './index.vue',
    output: { filename: 'bundle.js' },
    resolve: {
        extensions: ['.ts', '.vue']
    },
    module: {
        rules: [
            { test: /\.vue$/, loader: 'vue-loader' },
            { test: /\.ts$/, loader: 'ts-loader', options: { appendTsSuffixTo: [/\.vue$/] } }
        ]
    } 
}
```

**Comment by CodeLookBook** [comment]
@johnnyreilly 
As I understand I need to give more information. So, I use:
1) Laravel framework(use 'WebPack', 'laravel-mix' - that gives default webpack setting and opportunity to modify them)
2) VueJs 

About typescript+vuejs configuration:
I used this instructions to configure my VueJs project: https://github.com/Microsoft/TypeScript-Vue-Starter . According to them I installed next NPM packages: 
- typescript
- ts-loader
- and created file: 'vue-shims.d.ts.'


My 'tsconfig.json' file settings:

```
{
  "compilerOptions": {
    "sourceMap": true,
    "strict": true,
    "noImplicitReturns": true,
    "module": "es2015",
    "moduleResolution": "node",
    "target": "es5"
  },
  "include": [
    "./resources/**/*"   //Folder with my '.ts' and '.vue' files.
  ]
}
```

My web pack settings:
```
mix.webpackConfig({
    module: {
        rules: [
            {
                test: /\.tsx?$/,
                loader: 'ts-loader',
                exclude: /node_modules/,
                options: {
                    appendTsSuffixTo: [/\.vue$/],
                },
            },
        ],
    },
    resolve: {
        extensions: ['.ts', '.js', '.vue', '.json'],
        alias: {
            'vue$': 'vue/dist/vue.esm.js',
        },
    },
});
```

**Comment by johnnyreilly** [comment]
You appear to be missing vue-loader?

**Comment by CodeLookBook** [comment]
No. Laravel framework install a lot of packages automatically, among them: 
- vue-loader;
- vue-template-compiler;
- vue-template-es2015-compiler

**Comment by CodeLookBook** [comment]
@johnnyreilly 

No. Laravel framework install a lot of packages automatically, among them:

vue-loader;
vue-template-compiler;
vue-template-es2015-compiler

**Comment by johnnyreilly** [comment]
You are missing the vue-loader `rule` in your config. Add that and things may start to work

**Comment by johnnyreilly** [comment]
Closing as assumed resolved...

**Comment by CodeLookBook** [comment]
Your advice with using vue-loader `rule` didn't help to me. As I understood there is some strange sort of bug. I didn't notice that console gives me right messages about warnings and errors referenced with "interface" usage. But there is a problem with warnings and errors highlighting in "*.vue" files.
Examples:
![image](https://user-images.githubusercontent.com/12842434/27520284-2366b0f4-5a10-11e7-9dfa-830efe5714ce.png)


As you can see there is an error in code, and there is no any highlighting. But console. gave to me right message.

Also I  found that TypeScript doesn't see variables initialized inside `data(){return {...}} `(VueJS )

![image](https://user-images.githubusercontent.com/12842434/27520296-91834926-5a10-11e7-9bc9-1702bb309aed.png)

This time I'v got highlights and massage in the console but code was correct and there was no mistakes. And the most strange thing in this situation is that code was working fine with this errors messages )).


P.S: Thank you for your time.

**Comment by johnnyreilly** [comment]
I'm afraid I'm not a Vue user and so it's tricky for me to advise. Perhaps we should have an example in our set of examples. I wonder if @HerringtonDarkholme could advise - they know more than I on this particular topic....

**Comment by HerringtonDarkholme** [comment]
@CodeLookBook  

ts-loader does not magically configure your editor to report errors. As long as the console correctly reports type errors, I think ts-loader does its job.

> Also I found that TypeScript doesn't see variables initialized inside data(){return {...}}(VueJS )

I guess you point a different version of VueJS from the one specified in  https://github.com/Microsoft/TypeScript-Vue-Starter. Laravel's starter pack does not cover the setup  in the Starter.

> TypeScript interface doesn't work as expected with '*.vue' files. it doesn't throw any "warning" when provided type is bad. Example form vue.js file

Note in your example nothing is exported. And ts-loader only resolves `.vue` lazily, that is, if you don't import the Vue file, ts-loader won't report error for it. Eager resolving vue file, or any other non-ts file, is not a goal of ts-loader, IMHO.

In summary, I think ts-loader works as intended here. If you want more sophisticated editor experience, https://github.com/octref/vetur might be worth a try.

**Comment by CodeLookBook** [comment]
@johnnyreilly  @HerringtonDarkholme 
Thank you for your help!

**Comment by johnnyreilly** [comment]
Thanks @HerringtonDarkholme! 🌷
